### PR TITLE
feat(event_processor): add forwarding event processor and integrate with optimizely

### PR DESCRIPTION
**Pull request by rashidsp**
Summary
-------

-  Add default event processor as forwarding_event_processor.
-  Integrate changes with optimizely.
- Add validator that ensures that event_processor implements process method.

Test plan
---------

- Added unit tests.

### CODE REVIEW

**Comment by aliabbasrizvi** [comment]
Is this ready for review? I am ignoring as the wip label and the title is indicating this is not ready.

File: optimizely/event/event_processor.py [hunk 1]
```diff
@@ -266,3 +266,47 @@ def stop(self):
...
+
+
+class ForwardingEventProcessor(BaseEventProcessor):
+  """
+  ForwardingEventProcessor serves as the default EventProcessor.
+
+  The ForwardingEventProcessor sends the LogEvent to EventDispatcher as soon as it is received.
+  """
+
+  def __init__(self, event_dispatcher, logger, notification_center=None):
```
Comment (aliabbasrizvi) [inline optimizely/event/event_processor.py]: `logger` should be optional and we should set it to `NoOpLogger` if not provided.

File: optimizely/event/event_processor.py [hunk 2]
```diff
@@ -266,3 +266,47 @@ def stop(self):
...
+    """ ForwardingEventProcessor init method to configure event dispatching.
+
+    Args:
+      event_dispatcher: Provides a dispatch_event method which if given a URL and params sends a request to it.
+      logger: Optional component which provides a log method to log messages. By default nothing would be logged.
```
Comment (aliabbasrizvi) [inline optimizely/event/event_processor.py]: You say it is optional, but it is not so in practice.

File: optimizely/event/event_processor.py [hunk 3]
```diff
@@ -266,3 +266,47 @@ def stop(self):
...
+      notification_center: Optional instance of notification_center.NotificationCenter.
+    """
+    self.event_dispatcher = event_dispatcher
+    self.logger = logger
+    self.notification_center = notification_center
```
Comment (aliabbasrizvi) [inline optimizely/event/event_processor.py]: Similar to feedback in https://github.com/optimizely/python-sdk/pull/204 shouldn't we validate that `notification_center` is an instance of `notification_center.NotificationCenter`

File: optimizely/exceptions.py [hunk 4]
```diff
@@ -43,7 +43,7 @@ class InvalidGroupException(Exception):
 
 
 class InvalidInputException(Exception):
-  """ Raised when provided datafile, event dispatcher, logger or error handler is invalid. """
+  """ Raised when provided datafile, event dispatcher, logger, event processor or error handler is invalid. """
```
Comment (aliabbasrizvi) [inline optimizely/exceptions.py]: nit. Year in header needs to be updated.

File: optimizely/helpers/validator.py [hunk 5]
```diff
@@ -72,6 +72,19 @@ def is_config_manager_valid(config_manager):
...
+  """ Given an event_processor, determine if it is valid or not i.e. provides a process method.
+
+  Args:
+    event_processor: Provides a process method to create user events and then send requests.
+
+  Returns:
+    Boolean depending upon whether event_processor is valid or not.
+  """
+
+  return _has_method(event_processor, 'process')
```
Comment (aliabbasrizvi) [inline optimizely/helpers/validator.py]: Shouldn't this check that it implements `BaseEventProcessor`?

Comment (oakbani) [reply optimizely/helpers/validator.py]: Doesn't it provides more flexibility to the implementer. We do the same for is_logger_valid, is_event_dispatcher_valid, is_error_handler_valid, is_config_manager_valid in the same class.

Comment (oakbani) [reply optimizely/helpers/validator.py]: Doesn't it provides more flexibility to the implementer. We do the same for is_logger_valid, is_event_dispatcher_valid, is_error_handler_valid, is_config_manager_valid in the same class.

Comment (aliabbasrizvi) [reply optimizely/helpers/validator.py]: Makes sense.

**Comment by oakbani** [comment]
@msohailhussain Have fixed a failing FSC test in this PR. 
- Now we only wait for 50ms, previously we were waiting for 100ms
- Now we also allow flush interval in seconds between 0 and 1.

**Comment by oakbani** [comment]
Latest Event Batching Travis Build with Event Batching Testapp Branch https://travis-ci.com/optimizely/fullstack-sdk-compatibility-suite/builds/130617595

**Comment by oakbani** [comment]
Should pass once https://github.com/optimizely/python-sdk/pull/211 is merged

File: tests/test_optimizely.py [hunk 6]
```diff
@@ -339,7 +362,9 @@ def on_track(event_key, user_id, attributes, event_tags, event):
         self.assertTrue(isinstance(attributes, dict))
       if event_tags is not None:
         self.assertTrue(isinstance(event_tags, dict))
-      self.assertTrue(isinstance(event, event_builder.Event))
+
+      # TODO: what should be done about passing dicts of class instances?
+      # self.assertTrue(isinstance(event, LogEvent))
```
Comment (aliabbasrizvi) [inline tests/test_optimizely.py]: Why is this commented? Can we remove the print statement in the next line?

Comment (oakbani) [reply tests/test_optimizely.py]: Done

**Pull request merged**